### ISSUE TITLE
Error rate with wildcards

### ISSUE THREAD

**Issue by dariober** [marcelm/cutadapt#585]
Using cutadapt 3.5 with Python 3.9.7 installed with conda.

Consider this `test.fq` file:

```
@r1
TTTAACCTTCGCGCCGCGCCCCGGCCCGGG
+
IIIIIIIIIIIIIIIIIIIIIIIIIIIIII
``` 

I want to trim the leading `TTTAACCTT` allowing 1 mismatch and using adapter `^NNNAAGCTTC`. There is one mismatch between read (`TTTAACCTT`) and adapter (`NNNAAGCTTC`) so I would expect cutadapt to trim it, but it doesn't:

```
cutadapt -g '^NNNAAGCTTC' --error-rate 1 --info-file test.info test.fq
...
Total reads processed:                       1
Reads with adapters:                         0 (0.0%)    <<<<<<
Reads written (passing filters):             1 (100.0%)
...
```

Allowing 2 errors works but the test.info file reports 1 mismatch, as I would expect:

```
cutadapt -g '^NNNAAGCTTC' --error-rate 2 --info-file test.info test.fq
...
Reads with adapters:                         1 (100.0%)
...

cat test.info
r1	1	0	10		TTTAACCTTC	GCGCCGCGCCCCGGCCCGGG	1		IIIIIIIIII	IIIIIIIIIIIIIIIIIIII
       ^^^
```

Does the wildcard sequence `NNN` count as 1 mismatch or am I missing something? Thanks!

**Comment by marcelm** [comment]
I’m not sure, yet, but this is probably a bug. The alignment algorithm always uses an error rate internally, so if you use `--error-rate 1`, this is divided by the length of the adapter sequence to get the maximum error rate (1 / 10 = 0.1 in the first case). However, when checking whether an adapter match does not exceed the allowed error rate, the division is actually (number of errors) / (number of non-N bases in the matching part of the adapter). In your case, that would be 1/7, which is greater than 0.1, and therefore no match is detected.

Probably this should be changed such that the number of errors provide via `--error-rate` is divided by the number of non-N bases in the adapter as well.

**Comment by marcelm** [comment]
> Does the wildcard sequence `NNN` count as 1 mismatch or am I missing something? Thanks!

And just to clarify: No, matches to `N` bases always incur zero errors (unless you use `--no-match-adapter-wildcards`).

**Comment by marcelm** [comment]
This has been fixed and the fix will be available in the next release (probably Cutadapt 3.8). See also #603.
